### PR TITLE
Support for sending and receiving UNIX_FDs

### PR DESCRIPTION
**Pull request by exvito**
Hello,

First things first: thanks for sharing txdbus.

I've been exploring bluetooth programming on Linux, which from an application standpoint is mostly done over dbus, as far as I can tell. Being very familiar with Twisted, it was great finding txdbus.

One of the challenges I found was the need to receive UNIX_FDs as arguments to methods implemented in my code as in `NewConnection` below:

```
_iface = interface.DBusInterface(
    'org.bluez.Profile1',
    interface.Method('NewConnection', arguments='oha{sv}', returns=''),
    interface.Method('RequestDisconnection', arguments='o', returns=''),
    interface.Method('Release', arguments='', returns=''),
    interface.Method('Cancel', arguments='', returns=''),
)
```

I found it odd for a process to send/receive an open FD over IPC/RPC mechanisms, but later found it not only useful but, at least, supported over UNIX domain socket transports. Interesting concept!

Then I did two things:
* I lighly studied the dbus spec at https://dbus.freedesktop.org/doc/dbus-specification.html.
* I found out that Twisted supports sending/receiving FDs on POSIX platforms via https://twistedmatrix.com/documents/current/api/twisted.internet.interfaces.IFileDescriptorReceiver.html

After that I set out to study txdbus code and came up with this first take. Highlights:
* It's delivering the UNIX_FD argument as an integer. Motive: the FD may represent an open file, socket, pipe, etc. It is up to the callee to know that and to optionally build a Python object from that.
* Clearly the other half is missing, supporting ```callRemote``` with UNIX_FD type specs and arguments.
* It probably needs a way of activating/deactivating UNIX_FD support in the authenticator - IIUC, dbus transports can be non-UNIX domain sockets over which, as far as I can tell, such capability is not supported - this take is hardcoding such support in ```ClientAuthenticator```.
* I believe that the ```self._receivedFDs``` needs to be reset after processing in ```rawDBusMessageReceived``` in the ```BasicDBusProtocol``` implementation.
* Three tests are failing for now: two related to authentication and one related to message parsing. I'll propose handling these once the items above are addressed.
* It is working for me in prototype code that gets an FD to an RFCOMM bluetooth socket with which I can send/receive data pretty much like a TCP STREAM oriented socket.

I'm looking for review and feedback, in particular regarding authentication -- not sure what's a client vs. a server and whether UNIX_FD support should automatically depend on transport or not -- and, of course, on the general approach at the code I'm sharing.

I have a few ideas on how to go about the other half but will standby on feedback.

Thanks in advance for any input.
Cheers.

PS: This starts addressing issue #24 even though for the reasons stated above, it is not delivering file objects.

### CODE REVIEW

**Comment by cocagne** [comment]
Overall, this looks great. Thanks exvito! I only have one main concern to address before accepting the pull request and that's the potential for breakage when non-unix domain transports are being used. Normally I'd also like to see a unit test written to validate the functionality and protect against future regressions but, in this case, I can't think of a particularly easy-to-implement test. Maybe launching a small test script that opens a pipe, connects to dbus, passes the read end of the pipe across, and then sends a "hello world" string? Give it some thought. If it's not too difficult to implement, it'd probably be worth your time to protect against future breakage. If not though, I'll take your word for it. Anyway, nice work on this. Check that one issue and think about a unit test and I'll accept the pull request. Thanks!

Oh, and a bit of background on this. File descriptor passing is a special feature of Unix-domain sockets. They have at least one other special feature I'm aware of and that's being able to retrieve the UID/GID of the process on the other end of the connection (kernel guarantees only root processes are capable of forgery). As far as I know, Unix sockets are the one and only mechanism for passing file descriptors on Unix platforms (which includes OSX). A major use case for fd-passing is network servers that have a micro process that accepts incoming connections and then passes the file descriptor off to a background process to handle the connection (and probably the motivating factor for adding this capability). Anyway, there isn't much of a client/server relationship here. Either side can send an fd across. The sender is the "server" and the receiver is the "client".

**Comment by exvito** [comment]
Thanks for your quick feedback and extra bit of background.

Here's what I think I'll do:
* Fix the authentication code - only negotiate UNIX_FD if transport is a UNIX domain socket.
* Adjust authentication tests to verify that.
* Implement sending UNIX_FDs as arguments to ```callRemote```.
* Create tests for passing around UNIX_FDs both as arguments and return values for both caller and callee.
* Review (with your guidance, if possible) the need to update any relevant docs.

This will take some time and I'm not sure I'll be able to dedicate a lot of it till the end of the month -- but I'll certainly be able to share some progress in the meantime.

Once there, this PR may be in shape for merging -- with tests as you suggest and supporting both sending and receiving.

**Comment by exvito** [comment]
Here's some progress:
* UNIX_FD support now only negotiated over UNIX domain socket transports.
* Not sure about raising the excepion when receiving an unexpected "AGREE_UNIX_FD" (line 132 in authentication.py).
* Auth tests were updated - not 100% happy with the approach, but they seem to work.

**Comment by cocagne** [comment]
Sounds like a good plan. I'll try and stay up to date with your changes as
you make them. With respect to documentation, it'll probably be sufficient
to simply provide a client/server example; possibly derived from your unit
test.

**Comment by exvito** [comment]
I confess I couldn't let my mind go of this... I stretched my evening and came up with a first take at supporting sending UNIX_FDs as arguments to callRemote.

I also included a *very* simple example client and server that serve to demonstrate correct operation. I'd like to extend these to other cases like sending/receiving more than one UNIX_FD and also returning UNIX_FDs as results -- this will probably need more coding.

I had to bend my head a bit around the need to update the message headers. This leads me to believe that the receiving side of the code on the first commit should also verify those -- for now it is simply assuming and using FDs per method signatures and throwing them away after parsing.

Feedback is appreciated.

**Comment by cocagne** [comment]
This is looking really good and I think you're definitely on the right
track here. You mention that your example is "very" simple but I think you
could actually make it even simpler and more effective. It took me a minute
or two to wrap my head around what the intent of the examples were so I
think, at a minimum, a comment block that describes what the example is
trying to demonstrate would be a good thing (most people that find your
examples will probably only have a fuzzy idea about what this whole
file-descriptor passing thing is anyway... it's pretty esoteric). Beyond
that, you might consider having the client use os.pipe() to create a pair
of descriptors then pass the write end to the server and read back and
print  the "Hello World" string the server writes to the fd. Sure, it's a
blocking operation but the rest of Twisted blocks on file descriptor reads
anyway (... which there doesn't seem to be any good way around at present).

Anyway, nice work exvito :)

Tom

**Comment by exvito** [comment]
Thanks for you feedback, Tom.

I haven't (yet) gotten to updating the examples with you os.pipe suggestion, which I think will be a nice approach. Instead I played around a bit with the client/server to exercise the implementation.

Unfortunatelly, I found serious limitations, which I noted on a comment in fd_client.py:
* Implementation fails client-side when a second remote call with a UNIX_FD argument is attempted. The behavior is as follows:
    * First remote call completes successfully.
    * Second remote call never reaches the server.
    * DBus closes the connection with the caller.
    * The server keeps running and responds correctly to subsequent calls from the next client run.
* Implementation fails server-side when handling a remote call with two UNIX_FD arguments. The behavior is as follows:
    * Client sends call message to DBus.
    * DBus fowards it to server.
    * Server fails with a struct.error exception in twisted/internet/unix.py.
    * Server is disconnected from DBus (not shure who initiates the disconnection, server or dbus).
    * Client fails, as expected, with org.freedesktop.DBus.Error.NoReply.

I'm currently a bit at a loss. DBus logs showed nothing and not being very familiar with dbus-monitor, all I can say is that in the second case, the two UNIX_FDs seem to be going through and look fine.

I hacked a bit of twisted/internet/unix.py:doRead, where the exception is raised, and realised that dbus is sending the UNIX_FDs in a format that isn't expected by Twisted (something along the lines of Twisted seems to expect one UNIX_FD per message, DBus seems to be sending both).

Maybe this (yet to be confirmed) different aproach at sending/receiving UNIX_FDs between Twisted and DBus also explains the first failure.

Other than recompiling DBus with DBUS_VERBOSE support -- which the one I'm running on Ubuntu Server 16.04 does not seem to have -- and examining DBus source code for the UNIX_FD sending/receiving implementation, I'm not sure how to further diagnose this.

Do you have any ideas?
Thanks in advance.

**Comment by exvito** [comment]
Good news:
* Compiled DBus with DBUS_VERBOSE support -- much simpler than I anticipated.
* Ran with DBUS_VERBOSE=1 -- Wow, VERY VERBOSE!
* Found out the culprit for the first failure above: my implementation not building the message headers correctly.
* Fixed in the previous commit -- along with updated sample client.

PS: Not 100% happy with the header building approach, but it works for now. Feedback is welcome.

**Comment by exvito** [comment]
More good news / progress:
* I studied the dbus source code.
* Confirmed that it may send more than one UNIX_FD per message, while the current Twisted implementation assumes such messages contain a single UNIX_FD -- thus failing.
* Created a ticket https://twistedmatrix.com/trac/ticket/8911 and PR https://github.com/twisted/twisted/pull/599 for Twisted with a fix.

Next:
* Waiting on Twisted team to review my proposed change which may need improvements.

Stay tuned and/or chime in with any input.
Cheers.

**Comment by cocagne** [comment]
Sorry for the delayed response. Nice work! I wasn't sure it was possible to
send more than a single FD per message. And congrats on getting your patch
approved for merging into the Twisted mainline, those guys are pretty
rigorous in their reviews and what they accept :)

Once you know for sure which version of Twisted will pick up your changes,
you may want to consider adding a check on the Twisted version in the
UNIX_FD marshall/unmarshall code that detects the attempt to send/receive
multiple FDs in a single message and throws an explicit error stating that
this isn't supported until Twisted version X.Y.Z (pretty sure
twisted.version or twisted.__version__ or somesuch provides easy access to
the version being used).

Looking good,

Tom

**Comment by exvito** [comment]
> Once you know for sure which version of Twisted will pick up your changes,
> you may want to consider adding a check on the Twisted version in the
> UNIX_FD marshall/unmarshall code that detects the attempt to send/receive
> multiple FDs in a single message and throws an explicit error stating that
> this isn't supported until Twisted version X.Y.Z (pretty sure
> twisted.version or twisted.__version__ or somesuch provides easy access to
> the version being used).

I'm not sure such an approach will work. Twisted's APIs for sending/receiving UNIX_FDs are somewhat detached from the underlying calls to send/recvmsg and how many UNIX_FDs are carried in each such message. The actual APIs only support one UNIX_FD at a time. The current implementation (once my contribution gets in) sends one UNIX_FD per message but supports receiving more than one UNIX_FD per message (calling the protocol's receivedFileDescriptor as many times as UNIX_FDs received).

(apparently, dbus itself is happy to receive two UNIX_FDs in separate messages and forwarding them in a single message -- this is pretty much the test/example client/server I was exploring)

Having said this, from txdbus' prespective, using Twisted's API, there is no way to control how many UNIX_FDs are sent/received per message. All you know is that for versions < x.y.z (to be determined) Twisted will fail when receiving more than one UNIX_FD in a single message -- since it is on the receiving side of the stack, there is no way txdbus can handle that failure because it happens before txdbus gets to act.

In other words, the only safe way of dealing with this is requiring Twisted >= x.y.z in requirements.txt. Or maybe I'm mistaken and there is something else I'm not seeing clearly. :)

**Comment by cocagne** [comment]
Yeah, it may not work. Given that it's never been supported before at all,
it'd probably be sufficient to just note it in the README file on the git
repository and update the notes on pypi.

Tom

**Comment by exvito** [comment]
Tom,

I'm standing by on https://github.com/twisted/twisted/pull/599, which has been approved, to see when it will be merged and on which release it is included -- if it goes through, of course. :)

I think one additional possibility will be to log a warning if a server implements a method receiving more than one UNIX_FD while using a Twisted release earlier than x.y.z, to be determined -- this could guide developers to update Twisted if they need such capabilities.

Standing by. :)

**Comment by cocagne** [comment]
Yeah, that sounds like a good compromise. let me know when you're ready to
merge.

**Comment by cocagne** [comment]
ping

**Comment by exvito** [comment]
Hello Tom, thanks for the ping.

Here's a quick status:
* https://github.com/twisted/twisted/pull/599 was merged. It should be included in the Twisted release after 16.7.0, which is currently in rc1. Probably the first of the 17.x series.
* In the meantime, I filed https://github.com/twisted/twisted/pull/651, which fixes a  segmentation fault scenario, which I faced when exercising and testing these FD sending/receiving within Twisted. That was merged in, as well.
* Lastly, I have filed https://github.com/twisted/twisted/pull/652, which ensures some low-level validations, again improving robustness, by preventing data-integrity and improving the processing of incoming messages such as the ones holding UNIX_FDs. This one is still up for review.

As soon as I know the Twisted version in which the first PR is included in (the one that is actually required for txdbus), I'll proceed with completing this PR, if that's fine by you.

Regards.

**Comment by cocagne** [comment]
Sounds good. Go ahead and merge when you feel it's ready.

Tom

**Comment by exvito** [comment]
Tom,

After a somewhat long wait, Twisted 17.1.0 was released including the necessary fixes to handle multiple UNIX_FD receiving. I've updated the PR according to that:
* When declaring a method with more than one UNIX_FD argument under a Twisted version that does not support it, a RuntimeError is raised.
* Updated the example fd_client/fd_server.py.

NEXT:
* Planning on understanding test structure to include UNIX_FD passing tests.

As before, your feedback will be welcome.

**Comment by exvito** [comment]
@hobarrera, @cocagne,

I'm glad some activity has been going on, recently. I've had this PR on the backburner for a while and your recent actions prompted me to revisit it and push it through.

* The fist thing I did was merging in the updated master.
  * I had two easily fixed conflicts and was done with it.
* Then I ran `doc/examples/fd_server.py` + `doc/examples/fd_client.py` by hand:
  * Successfully on Python 2.7.
  * Failed on Python 3.5 -- fixed after two minor commits.

This seems to be almost ready for merging.

However, something is going on that I do not understand completely; for that, your guidance will be appreciated. A few facts I'm observing:

* In commit 029b160e2732239f9fcd47fd80e386881fd97de3 I included two tests for calling/implementing methods with FD arguments.
* I have been running them with `trial txdbus`, which results in:
  * Python 2.7 on my dev system: `Ran 258 tests in 1.918s` `PASSED (successes=258)`
  * Python 3.5 on my dev system: `Ran 166 tests in 0.217s` `FAILED (failures=5, errors=24, successes=140)` (more on this below)
* To my surprise, Travis failed precisely under 2.7 and succeeded in 3.5:
  * I then noticed that it is using the `runtests.py` script instead of `trial`.
  * More mysteriously, `runtests.py`:
    * Python 2.7 on my dev system: `Ran 259 tests in 1.792s` including my FD arg tests.
    * Python 2.7 on Travis: `Ran 164 tests in 0.137s` `FAILED (errors=8)`
    * Python 3.5 on my dev system: `Ran 90 tests in 0.040s` not including my FD arg tests.
    * Python 3.5 on Travis: `Ran 90 tests in 0.018s` not including my FD arg tests.

After a little bit of investigation here are a few questions:
* Why isn't `runtests.py` using `trial`, shouldn't it, given that, at least some tests are async?
* Travis does not seem to be running my FD arg tests. Maybe because it doesn't have a DBUS service running, is that it? If that is the case, how feasible would it be to start such a session for the test running procedure?

Wrapping up, I applied two minor Python 3 fixes to `txdbus/test/test_client_against_internal_bus.py` to at least make `trial` drive those tests. I also ensured that the FD arg tests run ok with `trial txdbus.test.test_client_against_native_bus.UnixFDArgumentsTester` under both Python 2 and 3 (see d8f5176aa046fafe7c0f543685344dd2ccc32705).

All in all, I'd love to wrap up my contribution and any related contributions (maybe around testing) so that txdbus gets both more robust (via testing and Travis) and more featureful (with this PR). For that, and to support merging this, maybe we can improve on testing, picking up on the Travis integration and recent contributions by @hobarrera -- which is looking very nice! :)

Thanks in advance.
Cheers.

**Comment by hobarrera** [comment]
Hi!

Thanks for the kind words and the contribution.

* `runtests.py` skips a lot of tests for python3, because they're still broken. There's a branch to work on these, but for now, we skip a lot.
* I didn't use `trial` merely because I didn't know of its existence. It sounds quite cool, but I'd leave it to *after* all the changes for py3 support are done.

Oh, and the tests that use the live DBus are only run *if* there's a DBus running, and it appears that travis does not run one (I'm not sure if we can run one or not, probably something to look into in future).

**Comment by exvito** [comment]
> runtests.py skips a lot of tests for python3, because they're still broken. There's a branch to work on these, but for now, we skip a lot.

Thanks for clarifying.

> I didn't use trial merely because I didn't know of its existence. It sounds quite cool, but I'd leave it to after all the changes for py3 support are done.

I just wonder if the async tests will work without trial... And, again, if so, I'm not sure I understand why (I'm no trial expert!)

> The tests that use the live DBus are only run if there's a DBus running, and it appears that travis does not run one (I'm not sure if we can run one or not, probably something to look into in future).

Googling a bit led me to https://github.com/Shouqun/node-dbus which seems to be:
* Installing the dbus package on some kind of Debian based system.
* Launching the with a `dbus-launch` thing... (FYI, I run `dbus-run-session <command>` on my dev system)

Disclaimer: I have zero experience setting up Travis!

Finally, what would your approach towards merging this PR be? Maybe waiting on Python 3 "formal" support and tests is the best approach. What would you say?

File: txdbus/interface.py [hunk 1]
```diff
@@ -34,6 +36,12 @@ def __init__(self, name, arguments='', returns='' ):
         self.nret    = -1
         self.sigIn  = arguments
         self.sigOut = returns
+        if arguments.count('h') > 1:
+            # More than one UNIX_FD argument requires Twisted >= 17.1.0.
+            minTxVersion = type(twisted.version)('twisted', 17, 1, 0)
+            if twisted.version < minTxVersion:
```
Comment (hobarrera) [inline txdbus/interface.py]: You can just use `twisted.version.base() < '17.1.0'` here.

**Comment by hobarrera** [comment]
> I just wonder if the async tests will work without trial... And, again, if so, I'm not sure I understand why (I'm no trial expert!)

They work, **but** occasionally, some tests fail (like, 2 out of every 10 runs), so there might be some race condition there.

> Googling a bit led me to https://github.com/Shouqun/node-dbus which seems to be:

Great stuff man! It also turns out that current travis images already have DBus installed, so it's just a matter of [adding the command](https://github.com/cocagne/txdbus/compare/master...hobarrera:dbus-on-travis).

I'm hesitant to merge that though, since native DBus tests fails from time to time (locally too).

> Finally, what would your approach towards merging this PR be? Maybe waiting on Python 3 "formal" support and tests is the best approach. What would you say?

I'm okay with merging it if tests pass for python2; don't let the py3 work hold you up. Final word is up to @cocagne, since I haven't really dug so much into the code to understand the full impact yet.

**Comment by hobarrera** [comment]
Actually, running `trail txdbus/test` works fine.

File: txdbus/interface.py (hunk 1 above)
Comment (exvito) [inline txdbus/interface.py]: Updated, per your suggestion.

**Comment by exvito** [comment]
> Actually, running trail txdbus/test works fine.

It does, but you should run `trial txdbus.test` instead (or `trial txdbus.test.test_authentication`, for example, for running a selection of the tests) -- for details, please check the trial docs/howto at http://twistedmatrix.com/documents/current/core/howto/trial.html

PS: If you're interested, I've got a quick-hack branch (almost equivalent to this PR) passing most of the 258 tests (<10 failing) under Python 3 here https://github.com/exvito/txdbus/tree/fd-support-py3

Regarding merging this PR, I guess I'll hold, unless @cocagne suggests otherwise. It somehow feels wrong knowing that Travis says "build failed". ;)

**Comment by exvito** [comment]
Following on the recent work on testing improvements I now feel this is pretty much ready for merging. This was ok with @cocagne back in January (see comment history).

If no one finds anything wrong during the next week, I suggest we merge it by then -- otherwise, as work progresses, it may become unnecessarily heavy to keep up with master.

Feedback welcome. :)

**Comment by hobarrera** [comment]
Can you please rebase to master (or let me know if you're not sure how and I can do it). If travis is happy after that, I'm happy to merge this.

**Comment by exvito** [comment]
Completed -- I actually had done that just a while ago -- that's why the tests were passing.

After updating with master, it's passing on Python 2 (see UnixFDArgumentsTester) but lots of things are failing under Python 3.

**Comment by exvito** [comment]
Now looking good. :)

**Comment by hobarrera** [comment]
Whoops, I missed your last commit -- I'd added it manually and merged already.

Thanks!

**Comment by hobarrera** [comment]
Double oops; master is [failing](https://travis-ci.org/cocagne/txdbus/jobs/238890830). Did I mismerge something?

**Comment by exvito** [comment]
You certainly did -- Commig 2d91d7c (my last one in this PR) passed all tests on all Python releases as you can see. :)

PS: Why didn't you merge from here? It would have been cleaner (in that github would track it clearly) and less error-prone.

PPS: txdbus/test/test_authentication.py **is not looking** good in `test_unix_fd_agree`... Compare with the commit I made -- from a quick visual inspection (not to be 100% trusted -- diff is your friend) you are missing a few b'' prefixes there.

**Comment by hobarrera** [comment]
My bad: I'd merged manually because I'd added a commit (turns out you'd already pushed **almost** the same).

**Comment by hobarrera** [comment]
Sorry for the mess, my bad -- I should have just waited for you. 😅  
Merged `master` into this branch to fix conflicts, will re-merged as soon as travis is happy.

**Comment by hobarrera** [comment]
I undid that mess I made -- can you re-push your last commit?

Again, I'm very sorry about the mess I made :(

**Comment by exvito** [comment]
@hobarrera,

I admit I felt you were a bit in a rush, yes... ;) I'll review and update the branch once I get it back to a merge-able state while ensuring all tests pass -- but I'm not sure I can do it before monday. 

Enjoy your weekend. :)

**Comment by hobarrera** [comment]
Yes. Sorry again about that. I'd too much on my plate today.

**Comment by cocagne** [comment]
Nice work guys! I'm a bit late in getting a chance to catch up on the
thread but I agree completely with your reasoning on the decision to merge.
One point of note with respect to the unit tests is that I would definitely
expect intermittent failures when using an external DBus daemon with a
non-trial based unit testing tool. The only real enhancement made by trial
is the ability to return a Deferred to the result of the test rather than
having it complete immediately. Given the way the code is structured, I can
believe that non-async based unit testing frameworks would generally either
"get lucky" where the tests would complete before the function returns or
it would provide false successes in that the method would complete almost
immediately and therefore show success but not actually run some or all of
the test code. So using anything other than trial is iffy at best unless
the testing framework explicitly includes support for a trial plugin.
pytest has that, I believe, but I'm not sure about others. Personally, I've
always just used trial.

Tom

**Comment by hobarrera** [comment]
Completely agree. We're using plain `trail` for now already (the way you'd originally written the tests was 100% compatible 😉 ), so there should be no such errors (hopefully!).

**Comment by exvito** [comment]
@hobarrera,

I just updated this PR from the current master, yet again fixing merge conflicts. Tests should pass now and I believe it would be a great opportunity to merge, before master changes again "under my feet", forcing me to rebase/re-merge master into this branch.

Of course, your review is welcome. :)

**Comment by exvito** [comment]
Yep, tests passed. :)

**Comment by exvito** [comment]
Thanks @hobarrera + @cocagne. :)
